test/acceptance/js/FilestoreApp.js

```diff
@@ -1,109 +1,112 @@
+  constructor() {
+    this.running = false
+    this.initing = false
+  }
+
+  async runServer() {
     if (this.running) {
-      return callback()
-    } else if (this.initing) {
-      return this.callbacks.push(callback)
-    } else {
-      this.initing = true
-      this.callbacks.push(callback)
-      return app.listen(
-        __guard__(
-          Settings.internal != null ? Settings.internal.filestore : undefined,
-          x => x.port
-        ),
+      return
+    }
+
+    if (this.initing) {
+      return this.waitForInit()
+    }
+    this.initing = true
+
+    this.app = await FilestoreApp.requireApp()
+
+    await new Promise((resolve, reject) => {
+      this.server = this.app.listen(
+        Settings.internal.filestore.port,
         'localhost',
-        error => {
-          if (error != null) {
-            throw error
+        err => {
+          if (err) {
+            return reject(err)
           }
-          this.running = true
-          logger.log('filestore running in dev mode')
-
-          return (() => {
-            const result = []
-            for (callback of Array.from(this.callbacks)) {
-              result.push(callback())
-            }
-            return result
-          })()
+          resolve()
         }
       )
+    })
+
+    if (Settings.filestore.backend === 's3') {
+      try {
+        await FilestoreApp.waitForS3()
+      } catch (err) {
+        await this.stop()
+        throw err
+      }
+    }
+
+    this.initing = false
+  }
+
+  async waitForInit() {
+    while (this.initing) {
+      await new Promise(resolve => setTimeout(resolve, 1000))
     }
-  },
-
-  waitForS3(callback, tries) {
-    if (
-      !(Settings.filestore.s3 != null
-        ? Settings.filestore.s3.endpoint
-        : undefined)
-    ) {
-      return callback()
+  }
+
+  async stop() {
+    if (this.server) {
+      await new Promise(resolve => {
+        this.server.close(resolve)
+      })
+      delete this.server
     }
-    if (!tries) {
-      tries = 1
+  }
+
+  static async waitForS3() {
+    let tries = 0
+    if (!Settings.filestore.s3.endpoint) {
+      return
     }
 
-    return request.get(
-      `${Settings.filestore.s3.endpoint}/`,
-      (err, response) => {
-        console.log(
-          err,
-          response != null ? response.statusCode : undefined,
-          tries
+    let s3Available = false
+
+    while (tries < S3_TRIES && !s3Available) {
+      try {
+        const response = await promisify(request.get)(
+          `${Settings.filestore.s3.endpoint}/`
         )
-        if (
-          !err &&
-          [200, 404].includes(
-            response != null ? response.statusCode : undefined
-          )
-        ) {
-          return callback()
+        if ([200, 404].includes(response.statusCode)) {
+          s3Available = true
         }
-
-        if (tries === S3_TRIES) {
-          return callback('timed out waiting for S3')
+      } catch (err) {
+      } finally {
```

Comment:
I would add a comment in the `catch` clause indicating that errors are all swallowed on purpose.